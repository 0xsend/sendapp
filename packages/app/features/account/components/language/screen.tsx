import { Card, Paragraph, Separator, Spinner, XStack, YStack } from '@my/ui'
import { SettingsHeader } from 'app/features/account/components/SettingsHeader'
import { useTranslation } from 'react-i18next'
import { useCallback, useMemo, useState } from 'react'
import { Pressable, Platform } from 'react-native'
import { Check } from '@tamagui/lucide-icons'
import { getSupportedLocaleOptions } from 'app/i18n/locales'
import { useHoverStyles } from 'app/utils/useHoverStyles'

export const LanguagePreferences = () => {
  const { i18n, t } = useTranslation('settings')
  const [pendingLocale, setPendingLocale] = useState<string | null>(null)
  const resolvedLanguage = i18n.resolvedLanguage ?? i18n.language

  const options = useMemo(() => getSupportedLocaleOptions(), [])
  const hoverStyles = useHoverStyles()

  const handleSelect = useCallback(
    async (locale: string) => {
      if (locale === resolvedLanguage || pendingLocale === locale) {
        return
      }

      setPendingLocale(locale)
      try {
        await i18n.changeLanguage(locale)
      } finally {
        setPendingLocale(null)
      }
    },
    [i18n, pendingLocale, resolvedLanguage]
  )

  return (
    <YStack gap={'$3.5'} w={'100%'}>
      {Platform.OS === 'web' && <SettingsHeader>{t('language.title')}</SettingsHeader>}
      <Card gap={'$3'} padded size={'$4'}>
        <Paragraph size={'$4'} color={'$color12'}>
          {t('language.description')}
        </Paragraph>
        {pendingLocale ? (
          <Paragraph size={'$2'} color={'$lightGrayTextField'}>
            {t('language.updating')}
          </Paragraph>
        ) : null}
        <Separator boc={'$darkGrayTextField'} opacity={0.2} />
        <YStack>
          {options.map((option, index) => {
            const isActive = option.value === resolvedLanguage
            const isPending = pendingLocale === option.value

            return (
              <YStack key={option.value}>
                <Pressable
                  accessibilityRole="button"
                  accessibilityState={{ selected: isActive, busy: isPending }}
                  disabled={isPending}
                  onPress={() => handleSelect(option.value)}
                >
                  <XStack
                    ai="center"
                    jc="space-between"
                    px={'$2'}
                    py={'$3'}
                    br={'$4'}
                    hoverStyle={hoverStyles}
                  >
                    <Paragraph size={'$5'}>{option.autonym}</Paragraph>
                    {isPending ? (
                      <Spinner size="small" />
                    ) : isActive ? (
                      <Check size={16} color={'$primary'} />
                    ) : null}
                  </XStack>
                </Pressable>
                {index < options.length - 1 && (
                  <Separator boc={'$darkGrayTextField'} opacity={0.15} />
                )}
              </YStack>
            )
          })}
        </YStack>
      </Card>
    </YStack>
  )
}
