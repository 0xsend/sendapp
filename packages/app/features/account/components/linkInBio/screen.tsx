import {
  FadeCard,
  Paragraph,
  SubmitButton,
  Text,
  YStack,
  XStack,
  Input,
  Spinner,
  useMedia,
} from '@my/ui'
import { useCallback, useEffect, useState } from 'react'
import { FormProvider, useForm } from 'react-hook-form'
import { SettingsHeader } from 'app/features/account/components/SettingsHeader'
import { Platform } from 'react-native'
import {
  LinkInBioDomainNamesEnum,
  useLinkInBioMutation,
  type LinkInBioFormData,
} from 'app/utils/useLinkInBioMutation'
import { useUser } from 'app/utils/useUser'
import { IconLinkInBio } from 'app/components/icons'
import type { Database } from '@my/supabase/database-generated.types'

type LinkInBioDomainName = Database['public']['Enums']['link_in_bio_domain_names']

enum FormState {
  Overview = 'Overview',
  LinkInBioForm = 'LinkInBioForm',
}

export const LinkInBioScreen = () => {
  const media = useMedia()
  const { linksInBio, isLoading } = useUser()
  const form = useForm<LinkInBioFormData>()
  const { mutateAsync: updateLinkInBio } = useLinkInBioMutation()
  const [formState, setFormState] = useState<FormState>(FormState.Overview)
  const [errorMessage, setErrorMessage] = useState<string | null>(null)
  const [domainHandles, setDomainHandles] = useState<Record<LinkInBioDomainName, string>>(
    {} as Record<LinkInBioDomainName, string>
  )

  const handleSubmit = useCallback(async () => {
    setErrorMessage(null)

    try {
      // Convert domain handles to LinkInBioFormData format
      const linkInBioData: LinkInBioFormData = Object.entries(domainHandles || {}).map(
        ([domain_name, handle]) => ({
          domain_name: domain_name as LinkInBioDomainName,
          handle: handle.trim(),
        })
      )

      await updateLinkInBio(linkInBioData)
      setFormState(FormState.Overview)
    } catch (error) {
      console.error(error)
      if (error?.message) {
        setErrorMessage(error.message)
      }
    }
  }, [domainHandles, updateLinkInBio])

  useEffect(() => {
    if (linksInBio && Array.isArray(linksInBio)) {
      const handles: Record<LinkInBioDomainName, string> = {} as Record<LinkInBioDomainName, string>
      for (const link of linksInBio) {
        handles[link.domain_name] = link.handle ?? ''
      }
      setDomainHandles(handles)
    }
  }, [linksInBio])

  const updateDomainHandle = useCallback((domain_name: string, handle: string) => {
    setDomainHandles((prev) => ({
      ...prev,
      [domain_name]: handle,
    }))
  }, [])

  const renderAfterContent = useCallback(
    ({ submit }: { submit: () => void }) => (
      <YStack>
        <SubmitButton onPress={() => submit()}>
          <SubmitButton.Text>SAVE CHANGES</SubmitButton.Text>
        </SubmitButton>
        {errorMessage && (
          <Paragraph marginTop={'$5'} theme="red" color="$color9">
            {errorMessage}
          </Paragraph>
        )}
      </YStack>
    ),
    [errorMessage]
  )

  const socialLinksForm = (
    <YStack gap={'$5'}>
      <FadeCard elevation={'$0.75'}>
        {errorMessage && (
          <Paragraph marginBottom={'$5'} theme="red" color="$color9">
            {errorMessage}
          </Paragraph>
        )}
        <YStack gap={'$4'}>
          {Object.values(LinkInBioDomainNamesEnum).map((domain_name) => {
            const currentHandle = domainHandles[domain_name] || ''

            return (
              <YStack key={domain_name} gap={'$2'}>
                <XStack gap={'$4'} alignItems="center">
                  <IconLinkInBio domain_name={domain_name} />
                  <YStack flex={1} bbc={'$color12'}>
                    <Input
                      placeholder={domain_name}
                      placeholderTextColor={'$color10'}
                      value={currentHandle}
                      onChangeText={(value) => updateDomainHandle(domain_name, value)}
                      autoCapitalize="none"
                      autoCorrect={false}
                      bc={'$color0'}
                    />
                  </YStack>
                </XStack>
              </YStack>
            )
          })}
        </YStack>
      </FadeCard>
      {renderAfterContent({ submit: handleSubmit })}
    </YStack>
  )

  const overview = (
    <YStack gap={'$5'}>
      <FadeCard>
        {linksInBio && Array.isArray(linksInBio) && linksInBio.length > 0 ? (
          linksInBio.map(({ domain, handle, domain_name }, index) => {
            return (
              <YStack key={domain || `${domain_name}-${index}`}>
                <XStack gap={'$5'} ai="center">
                  <IconLinkInBio domain_name={domain_name} />
                  <Paragraph size={media.gtXs ? '$6' : '$4'} color={'$color12'}>
                    {`${domain}${handle ?? ''}`}
                  </Paragraph>
                </XStack>
              </YStack>
            )
          })
        ) : (
          <Text>No social links added yet.</Text>
        )}
      </FadeCard>
      <SubmitButton onPress={() => setFormState(FormState.LinkInBioForm)}>
        <SubmitButton.Text>edit links</SubmitButton.Text>
      </SubmitButton>
    </YStack>
  )

  if (isLoading) {
    return <Spinner size="small" color="$color12" />
  }

  return (
    <YStack w={'100%'}>
      <YStack gap={'$3.5'}>
        {Platform.OS === 'web' && <SettingsHeader>Link In Bio</SettingsHeader>}
        <FormProvider {...form}>
          {(() => {
            switch (formState) {
              case FormState.Overview:
                return overview
              case FormState.LinkInBioForm:
                return socialLinksForm
              default:
                return overview
            }
          })()}
        </FormProvider>
      </YStack>
    </YStack>
  )
}
