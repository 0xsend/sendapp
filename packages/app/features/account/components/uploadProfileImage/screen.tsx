import { SizableText, YStack } from '@my/ui'
import { IconRefresh } from 'app/components/icons'
import { useSupabase } from 'app/utils/supabase/useSupabase'
import { useUser } from 'app/utils/useUser'
import { decode } from 'base64-arraybuffer'
import * as ImagePicker from 'expo-image-picker'
import React, { Ref, forwardRef, useImperativeHandle, useState } from 'react'

export interface UploadAvatarRefObject {
  pickImage: () => void
}

export const UploadAvatar = forwardRef(function UploadAvatar(
  { children }: { children: React.ReactNode },
  ref: Ref<UploadAvatarRefObject>
) {
  const { user, updateProfile } = useUser()
  const supabase = useSupabase()
  const [errMsg, setErrMsg] = useState('')

  useImperativeHandle(ref, () => ({ pickImage }))

  const pickImage = async () => {
    // No permissions request is necessary for launching the image library
    const result = await ImagePicker.launchImageLibraryAsync({
      mediaTypes: ImagePicker.MediaTypeOptions.Images,
      allowsEditing: true,
      aspect: [1, 1],
      quality: 1,
      base64: true,
    })

    await uploadImage(result)
  }

  const uploadImage = async (pickerResult: ImagePicker.ImagePickerResult) => {
    if (pickerResult.canceled) {
      return
    }
    if (!user) return
    const image = pickerResult.assets[0]
    if (!image) {
      setErrMsg('No image provided.')
      return
    }

    const base64Image = image.base64

    if (!base64Image) {
      setErrMsg('No image provided.')
      return
    }

    const base64Str = base64Image.includes('base64,')
      ? base64Image.substring(base64Image.indexOf('base64,') + 'base64,'.length)
      : base64Image
    const res = decode(base64Str)

    if (!(res.byteLength > 0)) {
      setErrMsg('ArrayBuffer is null')
      // console.error('ArrayBuffer is null')
      return null
    }
    const result = await supabase.storage
      .from('avatars')
      .upload(`${user.id}/${Number(new Date())}.jpeg`, res, {
        contentType: 'image/jpeg',
        upsert: true,
      })
    if (result.error) {
      setErrMsg(result.error.message)
      return
      // console.log(result.error)
      // throw new Error(result.error.message)
    }
    const publicUrlRes = await supabase.storage
      .from('avatars')
      .getPublicUrl(result.data.path.replace('avatars/', ''))

    const { error: update_error } = await supabase
      .from('profiles')
      .update({ avatar_url: publicUrlRes.data.publicUrl })
      .eq('id', user.id)
    if (update_error) {
      return
    }
    await updateProfile()
  }

  return (
    <YStack gap={'$4'}>
      <YStack
        position="relative"
        alignSelf="flex-start"
        flexShrink={1}
        onPress={() => pickImage()}
        cursor="pointer"
      >
        {children}
        <YStack
          position="absolute"
          left={0}
          right={0}
          top={0}
          bottom={0}
          jc="center"
          ai="center"
          zIndex={100}
        >
          <YStack
            backgroundColor="black"
            opacity={0.34}
            borderRadius="$3"
            position="absolute"
            left={0}
            right={0}
            top={0}
            bottom={0}
            $md={{ borderRadius: 12 }}
          />
          <YStack position="absolute" left={0} right={0} top={0} bottom={0} jc="center" ai="center">
            <IconRefresh color="primary" />
          </YStack>
        </YStack>
      </YStack>
      {errMsg ? <SizableText theme="red">{errMsg}</SizableText> : <></>}
    </YStack>
  )
})
