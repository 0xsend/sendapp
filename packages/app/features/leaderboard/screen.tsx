import { Avatar, Card, H2, H3, LinearGradient, Paragraph, XStack, YStack } from '@my/ui'
import { useLeaderboard } from './utils/useLeaderBoard'
import type { LeaderboardEntry } from 'app/utils/zod/leaderboard'

export function LeaderboardScreen() {
  const { data } = useLeaderboard()

  return (
    <YStack w={'100%'} gap={'$size.3.5'} pb={'$size.2'} $gtMd={{ pt: 0 }} pt={'$size.3.5'}>
      <YStack gap={'$size.0.9'}>
        <H2 tt={'uppercase'} fontWeight={'900'} testID="mainTitle">
          Best in Class
        </H2>
        <Paragraph color={'$color10'} size={'$5'}>
          Register a Sendtag, maintain the minimum balance, and refer others to rise in the ranks.
        </Paragraph>
      </YStack>
      {data?.referrals.length ? (
        <YStack
          gap={'$size.1.5'}
          $gtMd={{ flexDirection: 'row' }}
          // @TODO: remove when we get transactions
          maxWidth={'600px'}
        >
          <Leaderboard title="Referrals" list={data?.referrals} />
          {/* <Leaderboard title="Transactions" list={data?.rewards} /> */}
        </YStack>
      ) : (
        <Paragraph>Unable to find leaderboard data</Paragraph>
      )}
    </YStack>
  )
}

function Leaderboard({ title, list }: { title: string; list: LeaderboardEntry[] }) {
  const isReferrals = title === 'Referrals'
  return (
    <Card f={1} pb={0}>
      <YStack br="$8">
        <YStack gap="$size.0.9" p="$5" $gtMd={{ p: '$8', pb: '$size.0.9' }} pb="$size.0.9" br="$8">
          <H3 pb={'$size.0.9'} fontWeight={'600'} size={'$7'} testID={`title${title}`}>
            {title}
          </H3>
          <LeaderBoardHeader isReferrals={isReferrals} />
          <LeaderboardList list={list} isReferrals={isReferrals} />
        </YStack>
        <LinearGradient
          pos="absolute"
          w="100%"
          h={'$size.9'}
          locations={[0, 1]}
          colors={['transparent', 'black']}
          b={0}
          borderBottomRightRadius={'$8'}
          borderBottomLeftRadius={'$8'}
        />
      </YStack>
    </Card>
  )
}

function LeaderBoardHeader({ isReferrals }: { isReferrals: boolean }) {
  return (
    <XStack
      gap="$3"
      ai="center"
      jc="space-between"
      borderBottomWidth={2}
      borderBottomColor={'$color9'}
      pb="$size.0.9"
    >
      <XStack>
        <Paragraph fontFamily={'$mono'} color={'$color10'} tt={'uppercase'} w="$size.3">
          #
        </Paragraph>
        <Paragraph fontFamily={'$mono'} color={'$color10'} tt={'uppercase'}>
          Sendtag
        </Paragraph>
      </XStack>

      <Paragraph fontFamily={'$mono'} color={'$color10'} tt={'uppercase'} ta="right">
        {isReferrals ? 'Referrals' : 'Total transactions'}
      </Paragraph>
    </XStack>
  )
}

const getName = (user: LeaderboardEntry['user']) => {
  switch (true) {
    case Array.isArray(user.tags) && !!user.tags.length:
      return `/${user.tags[0]}`
    case !!user.name:
      return user.name
    case !!user.send_id:
      return user.send_id
  }
}

function LeaderboardList({
  list,
  isReferrals,
}: { isReferrals: boolean; list: LeaderboardEntry[] }) {
  return list.map(({ user, referrals, rewards_usdc }, i) => (
    <XStack
      gap={'$size.0.9'}
      ai="center"
      jc={'space-between'}
      key={user.send_id}
      testID={`${user.id}-${user.send_id}`}
    >
      <XStack gap={'$2'} flexShrink={1}>
        <Paragraph mb="0" w="$size.3" lineHeight="$4" color={'$color10'} fontFamily={'$mono'}>
          {i + 1}
        </Paragraph>
        <XStack f={1} gap={'$size.0.9'} ai={'center'}>
          <Avatar size="$2" borderRadius={'$3'}>
            <Avatar.Image src={user.avatar_url} />
            <Avatar.Fallback
              backgroundColor={'$decay'}
              $theme-light={{ backgroundColor: '$white' }}
            />
          </Avatar>
          <Paragraph mb="0" lineHeight="$4" color="$primary">
            {getName(user)}
          </Paragraph>
        </XStack>
      </XStack>

      <Paragraph flexShrink={0} ta="right">
        {isReferrals ? referrals : rewards_usdc}
      </Paragraph>
    </XStack>
  ))
}
