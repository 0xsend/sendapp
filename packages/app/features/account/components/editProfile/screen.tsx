import {
  Button,
  Checkbox,
  FadeCard,
  Paragraph,
  ProfileAvatar,
  Separator,
  Spinner,
  SubmitButton,
  XStack,
  YStack,
} from '@my/ui'
import { SchemaForm } from 'app/utils/SchemaForm'
import { ProfileSchema, useProfileMutation } from 'app/utils/useProfileMutation'
import { useUser } from 'app/utils/useUser'
import { UploadAvatar, type UploadAvatarRefObject } from '../uploadProfileImage/screen'
import { useRef, useState } from 'react'
import type { Tables } from '@my/supabase/database.types'
import { Check } from '@tamagui/lucide-icons'
import { SettingsHeader } from 'app/features/account/settings/components/SettingsHeader'
import { ReadOnlyFieldWithLabel } from 'app/features/account/settings/components/ReadOnlyFieldWithLabel'
import { FieldWithLabel } from 'app/features/account/settings/components/FieldWithLabel'
import { useForm } from 'react-hook-form'
import type { z } from 'zod'

enum FormState {
  Overview = 'Overview',
  ProfileForm = 'ProfileForm',
}

export const EditProfile = () => {
  const [formState, setFormState] = useState<FormState>(FormState.Overview)
  const { profile } = useUser()

  return (
    <YStack w={'100%'}>
      <YStack gap={'$3.5'}>
        <SettingsHeader>Profile</SettingsHeader>
        {(() => {
          switch (true) {
            case !profile:
              return <Spinner size="large" />
            case formState === FormState.Overview:
              return (
                <Overview profile={profile} onPress={() => setFormState(FormState.ProfileForm)} />
              )
            case formState === FormState.ProfileForm:
              return (
                <EditProfileForm
                  profile={profile}
                  onSave={() => setFormState(FormState.Overview)}
                />
              )
          }
        })()}
      </YStack>
    </YStack>
  )
}

const Overview = ({ profile, onPress }: { profile: Tables<'profiles'>; onPress: () => void }) => {
  const { name, about, is_public, avatar_url } = profile
  const avatarRef = useRef<UploadAvatarRefObject>(null)

  return (
    <YStack gap={'$5'}>
      <FadeCard>
        <XStack gap={'$5'} width={'100%'}>
          <UploadAvatar ref={avatarRef}>
            <ProfileAvatar
              avatarUrl={avatar_url ? avatar_url : undefined}
              $gtMd={{ size: 88 }}
              size={88}
            />
          </UploadAvatar>
          <YStack gap={'$2'} ai={'flex-start'}>
            <YStack>
              <Paragraph
                size={'$5'}
                fontWeight={500}
                color={'$lightGrayTextField'}
                $theme-light={{ color: '$darkGrayTextField' }}
              >
                Profile Picture
              </Paragraph>
            </YStack>
            <Button unstyled onPress={() => avatarRef.current?.pickImage()}>
              <Button.Text
                textDecorationLine="underline"
                color="$primary"
                $theme-light={{ color: '$color12' }}
                size={'$5'}
              >
                Update
              </Button.Text>
            </Button>
          </YStack>
        </XStack>
        <FieldWithLabel label={'Name'}>
          <Paragraph size={'$8'} fontWeight={'500'}>
            {name || '-'}
          </Paragraph>
        </FieldWithLabel>
        <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
        <ReadOnlyFieldWithLabel label={'About'} text={about || '-'} />
        <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
        <XStack gap={'$3'} ai={'center'}>
          <Checkbox
            disabled={true}
            checked={!!is_public}
            borderWidth={0}
            backgroundColor={is_public ? '$primary' : '$background'}
            circular={true}
          >
            <Checkbox.Indicator>
              <Check color={'$black'} />
            </Checkbox.Indicator>
          </Checkbox>
          <Paragraph size={'$5'}>{`${is_public ? 'Public' : 'Private'}`} Profile</Paragraph>
        </XStack>
      </FadeCard>
      <SubmitButton theme="green" borderRadius={'$4'} p={'$4'} onPress={onPress}>
        <Button.Text ff={'$mono'} fontWeight={'500'} tt="uppercase" size={'$5'} color={'$black'}>
          edit profile
        </Button.Text>
      </SubmitButton>
    </YStack>
  )
}

function EditProfileForm({ profile, onSave }: { profile: Tables<'profiles'>; onSave: () => void }) {
  const { id, name, about, is_public } = profile
  const { mutateAsync, error } = useProfileMutation(id)
  const form = useForm<z.infer<typeof ProfileSchema>>()

  const handleSubmit = async () => {
    const values = form.getValues()
    await mutateAsync(values)
    onSave()
  }

  return (
    <SchemaForm
      form={form}
      schema={ProfileSchema}
      props={{
        name: {
          'aria-label': 'Name',
          bc: '$color0',
        },
        about: {
          'aria-label': 'Bio',
          placeholder: 'Tell us about yourself',
          backgroundColor: '$color0',
          rows: 2,
          focusStyle: {
            fontStyle: 'normal',
          },
        },
        isPublic: {
          defaultChecked: is_public !== null ? is_public : true,
        },
      }}
      defaultValues={{
        name: name ? name : '',
        about: about ? about : '',
        isPublic: is_public !== null ? is_public : true,
      }}
      formProps={{
        $gtSm: {
          maxWidth: '100%',
        },
      }}
      onSubmit={handleSubmit}
      renderAfter={({ submit }) => (
        <YStack>
          <SubmitButton
            theme="green"
            borderRadius={'$4'}
            p={'$4'}
            mt={'$1'}
            onPress={() => submit()}
          >
            <Button.Text
              ff={'$mono'}
              fontWeight={'500'}
              tt="uppercase"
              size={'$5'}
              color={'$black'}
            >
              SAVE CHANGES
            </Button.Text>
          </SubmitButton>
          {error && (
            <Paragraph marginTop={'$3'} theme="red" color="$color9">
              {error.message}
            </Paragraph>
          )}
        </YStack>
      )}
    >
      {({ name, about, isPublic }) => (
        <FadeCard>
          <FieldWithLabel label={'Name'} gap={'$2'}>
            {name}
          </FieldWithLabel>
          <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
          <FieldWithLabel label={'About'} gap={'$2'}>
            {about}
          </FieldWithLabel>
          <Separator boc={'$silverChalice'} $theme-light={{ boc: '$darkGrayTextField' }} />
          <XStack gap={'$3'} ai={'center'}>
            {isPublic}
            <Paragraph size={'$5'}>Make Profile Public</Paragraph>
          </XStack>
        </FadeCard>
      )}
    </SchemaForm>
  )
}
