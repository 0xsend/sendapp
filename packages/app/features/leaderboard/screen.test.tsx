import { expect } from '@jest/globals'
import { Provider } from 'app/__mocks__/app/provider'
import { act, render, screen } from '@testing-library/react-native'
import { LeaderboardScreen } from './screen'

describe('LeaderboardScreen', () => {
  it('renders leaderboard screen', async () => {
    jest.useFakeTimers()
    render(
      <Provider>
        <LeaderboardScreen />
      </Provider>
    )
    await act(async () => {
      jest.runAllTimers()
    })

    const title = await screen.findByTestId('mainTitle')
    expect(title).toHaveTextContent('Best in Class')
    expect(screen.toJSON()).toMatchSnapshot('LeaderboardScreen')
  })
})
