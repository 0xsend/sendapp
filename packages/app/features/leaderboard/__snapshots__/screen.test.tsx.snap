// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeaderboardScreen renders leaderboard screen: LeaderboardScreen 1`] = `
<View
  style={
    {
      "flexDirection": "column",
      "gap": 40,
      "paddingBottom": 28,
      "paddingTop": 40,
      "width": "100%",
    }
  }
>
  <View
    style={
      {
        "flexDirection": "column",
        "gap": 16,
      }
    }
  >
    <Text
      accessibilityRole="header"
      style={
        {
          "color": "#FFFFFF",
          "fontFamily": "System",
          "fontSize": 30,
          "fontWeight": "900",
          "lineHeight": 34,
          "marginBottom": 0,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 0,
          "textTransform": "uppercase",
          "userSelect": "auto",
        }
      }
      suppressHighlighting={true}
      testID="mainTitle"
    >
      Best in Class
    </Text>
    <Text
      style={
        {
          "color": "#B2B2B2",
          "fontFamily": "System",
          "fontSize": 19.2,
          "fontWeight": "400",
          "lineHeight": 21,
          "userSelect": "auto",
        }
      }
      suppressHighlighting={true}
    >
      Register a Sendtag, maintain the minimum balance, and refer others to rise in the ranks.
    </Text>
  </View>
  <View
    style={
      {
        "flexDirection": "column",
        "gap": 24,
        "maxWidth": "600px",
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#111f22",
          "borderBottomLeftRadius": 16,
          "borderBottomRightRadius": 16,
          "borderTopLeftRadius": 16,
          "borderTopRightRadius": 16,
          "flex": 1,
          "flexDirection": "column",
          "paddingBottom": 0,
          "position": "relative",
        }
      }
    >
      <View
        style={
          {
            "borderBottomLeftRadius": 22,
            "borderBottomRightRadius": 22,
            "borderTopLeftRadius": 22,
            "borderTopRightRadius": 22,
            "flexDirection": "column",
          }
        }
      >
        <View
          style={
            {
              "borderBottomLeftRadius": 22,
              "borderBottomRightRadius": 22,
              "borderTopLeftRadius": 22,
              "borderTopRightRadius": 22,
              "flexDirection": "column",
              "gap": 16,
              "paddingBottom": 16,
              "paddingLeft": 24,
              "paddingRight": 24,
              "paddingTop": 24,
            }
          }
        >
          <Text
            accessibilityRole="header"
            style={
              {
                "color": "#FFFFFF",
                "fontFamily": "System",
                "fontSize": 20,
                "fontWeight": "700",
                "lineHeight": 24,
                "marginBottom": 0,
                "marginLeft": 0,
                "marginRight": 0,
                "marginTop": 0,
                "paddingBottom": 16,
                "textTransform": "none",
                "userSelect": "auto",
              }
            }
            suppressHighlighting={true}
            testID="titleReferrals"
          >
            Referrals
          </Text>
          <View
            style={
              {
                "alignItems": "center",
                "borderBottomColor": "#666666",
                "borderBottomWidth": 2,
                "borderStyle": "solid",
                "flexDirection": "row",
                "gap": 13,
                "justifyContent": "space-between",
                "paddingBottom": 16,
              }
            }
          >
            <View
              style={
                {
                  "flexDirection": "row",
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#B2B2B2",
                    "fontFamily": "System",
                    "fontSize": 16.8,
                    "fontWeight": "400",
                    "lineHeight": 19,
                    "textTransform": "uppercase",
                    "userSelect": "auto",
                    "width": 36,
                  }
                }
                suppressHighlighting={true}
              >
                #
              </Text>
              <Text
                style={
                  {
                    "color": "#B2B2B2",
                    "fontFamily": "System",
                    "fontSize": 16.8,
                    "fontWeight": "400",
                    "lineHeight": 19,
                    "textTransform": "uppercase",
                    "userSelect": "auto",
                  }
                }
                suppressHighlighting={true}
              >
                Sendtag
              </Text>
            </View>
            <Text
              style={
                {
                  "color": "#B2B2B2",
                  "fontFamily": "System",
                  "fontSize": 16.8,
                  "fontWeight": "400",
                  "lineHeight": 19,
                  "textAlign": "right",
                  "textTransform": "uppercase",
                  "userSelect": "auto",
                }
              }
              suppressHighlighting={true}
            >
              Referrals
            </Text>
          </View>
          <View
            style={
              {
                "alignItems": "center",
                "flexDirection": "row",
                "gap": 16,
                "justifyContent": "space-between",
              }
            }
            testID="1-47788"
          >
            <View
              style={
                {
                  "flexDirection": "row",
                  "flexShrink": 1,
                  "gap": 7,
                }
              }
            >
              <Text
                style={
                  {
                    "color": "#B2B2B2",
                    "fontFamily": "System",
                    "fontSize": 16.8,
                    "fontWeight": "400",
                    "lineHeight": 19,
                    "marginBottom": "0",
                    "userSelect": "auto",
                    "width": 36,
                  }
                }
                suppressHighlighting={true}
              >
                1
              </Text>
              <View
                style={
                  {
                    "alignItems": "center",
                    "flex": 1,
                    "flexDirection": "row",
                    "gap": 16,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "borderBottomLeftRadius": 7,
                      "borderBottomRightRadius": 7,
                      "borderTopLeftRadius": 7,
                      "borderTopRightRadius": 7,
                      "flexDirection": "column",
                      "height": 28,
                      "justifyContent": "center",
                      "maxHeight": 28,
                      "maxWidth": 28,
                      "minHeight": 28,
                      "minWidth": 28,
                      "overflow": "hidden",
                      "position": "relative",
                      "width": 28,
                    }
                  }
                >
                  <View
                    style={
                      {
                        "bottom": 0,
                        "flexDirection": "column",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "zIndex": 1,
                      }
                    }
                  >
                    <Image
                      fullscreen={true}
                      onError={[Function]}
                      onLoad={[Function]}
                      source={
                        {
                          "uri": "https://i.pravatar.cc/500?u=Frederick Bartell",
                        }
                      }
                      style={
                        {
                          "height": 28,
                          "width": 28,
                        }
                      }
                    />
                  </View>
                  <View
                    style={
                      {
                        "backgroundColor": "#3E4A3C",
                        "bottom": 0,
                        "flexDirection": "column",
                        "left": 0,
                        "position": "absolute",
                        "right": 0,
                        "top": 0,
                        "zIndex": 0,
                      }
                    }
                  />
                </View>
                <Text
                  style={
                    {
                      "color": "#40FB50",
                      "fontFamily": "System",
                      "fontSize": 16.8,
                      "fontWeight": "400",
                      "lineHeight": 19,
                      "marginBottom": "0",
                      "userSelect": "auto",
                    }
                  }
                  suppressHighlighting={true}
                >
                  Sophie Jenkins Jr.
                </Text>
              </View>
            </View>
            <Text
              style={
                {
                  "color": "#FFFFFF",
                  "flexShrink": 0,
                  "fontFamily": "System",
                  "fontSize": 16.8,
                  "fontWeight": "400",
                  "lineHeight": 19,
                  "textAlign": "right",
                  "userSelect": "auto",
                }
              }
              suppressHighlighting={true}
            >
              10
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
